### PR TITLE
Adjust grid size dynamically based on change in area error

### PR DESCRIPTION
**Pull request by adisidev**
~In the future, it may also be helpful to make the `fill_with_desnity` smaller once grid size goes above 512.~ (implemented already now).

This considerably speeds up most maps, as they do not require as fine-grain a resolution as 512 x 512. Fine-tuning `grid_factor` may be looked in the future.

### CODE REVIEW

File: src/inset_state/inset_state.cpp [hunk 1]
```diff
@@ -385,6 +385,33 @@ void InsetState::set_area_errors()
...
+      ? 2
+      : default_grid_factor;
+  max_area_errors_.push_back(curr_max_area_error);
+  if (
+    n_finished_integrations_ >= 2 &&
+    curr_max_area_error > max_area_errors_[n_finished_integrations_ - 1] &&
+    curr_max_area_error > max_area_errors_[n_finished_integrations_ - 2]) {
+
+    // Multiply grid size with factor
+    std::cout << "Adjusting grid size." << std::endl;
```
Comment (nihalzp) [inline src/inset_state/inset_state.cpp:403]: Can we change std::cout to std::cerr? We want to only use std::cout to output cartogram with -o flag.

File: src/inset_state/inset_state.cpp [hunk 2]
```diff
@@ -385,6 +385,33 @@ void InsetState::set_area_errors()
...
+    lx_ *= grid_factor;
+    ly_ *= grid_factor;
+
+    // Reallocate FFTW plans
+    ref_to_rho_init()->allocate(lx_, ly_);
+    ref_to_rho_ft()->allocate(lx_, ly_);
+    make_fftw_plans_for_rho();
+    std::cout << "New grid dimensions: " << lx_ << " " << ly_ << std::endl;
```
Comment (nihalzp) [inline src/inset_state/inset_state.cpp:411]: Same as before.

**Review by nihalzp — CHANGES_REQUESTED** [review]

File: src/inset_state/flatten_density.cpp [hunk 3]
```diff
@@ -168,9 +168,10 @@ void InsetState::flatten_density()
   double t = 0.0;
   double delta_t = 1e-2;  // Initial time step.
   unsigned int iter = 0;
+  unsigned int max_iter = 300;
 
   // Integrate
-  while (t < 1.0) {
+  while (t < 1.0 || iter > max_iter) {
```
[the remark this answers is not in the document]
Comment (adisidev) [reply src/inset_state/flatten_density.cpp:174]: Indeed, will fix.

File: src/inset_state/fill_with_density.cpp [hunk 4]
```diff
@@ -101,19 +106,19 @@ void InsetState::fill_with_density(bool plot_density)
...
+
+        //   // Highlight where intersection is present
+        //   std::cerr << "\nInvalid Geometry!" << std::endl;
+        //   std::cerr << "Intersection of Polygons/Holes/Geodivs" << std::endl;
+        //   std::cerr << "Y-coordinate: " << y << std::endl;
+        //   std::cerr << "Left X-coordinate: " << left_x << std::endl;
+        //   std::cerr << "Right X-coordinate: " << right_x << std::endl;
+        //   std::cerr << std::endl;
+        //   // _Exit(8026519);
+        // }
```
[the remark this answers is not in the document]
Comment (adisidev) [reply src/inset_state/fill_with_density.cpp:121]: Was facing same error as you earlier: will fix now.

File: src/inset_state/inset_state.cpp [hunk 5]
```diff
@@ -385,6 +385,33 @@ void InsetState::set_area_errors()
...
+  double curr_max_area_error = max_area_error().value;
+  unsigned int grid_factor =
+    (long_graticule_length > default_long_graticule_length)
+      ? 2
+      : default_grid_factor;
+  max_area_errors_.push_back(curr_max_area_error);
+  if (
+    n_finished_integrations_ >= 2 &&
+    curr_max_area_error > max_area_errors_[n_finished_integrations_ - 1] &&
+    curr_max_area_error > max_area_errors_[n_finished_integrations_ - 2]) {
```
[the remark this answers is not in the document]
Comment (adisidev) [reply src/inset_state/inset_state.cpp:400]: Indeed, shall change this.

File: src/inset_state/inset_state.cpp (hunk 1 above)
Comment (adisidev) [reply src/inset_state/inset_state.cpp:403]: Alright, will change.

**Comment by nihalzp** [comment]
The content of this PR has been merged to the 'merge-pr' branch.

**Pull request closed without merging**